File: src/UnwindMC.Tests/SourceTests/VcTools.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using UnwindMC.Tests.Helpers;

namespace UnwindMC.Tests.SourceTests
{
    static class VcTools
    {
        private static readonly string ClExePath;
        private static readonly string DumpBinPath;

        static VcTools()
        {
            var vswhereExePath = Path.Combine(
                Environment.GetEnvironmentVariable("ProgramFiles(x86)"),
                "Microsoft Visual Studio", "Installer", "vswhere.exe");
            var json = Run(null, vswhereExePath, "-format", "json");
            var vsPath = JsonConvert
                .DeserializeObject<JArray>(json)
                .First(o => o["installationVersion"].ToObject<string>().StartsWith("15."))
                ["installationPath"]
                .ToObject<string>();
            var toolsRootDir = Path.Combine(vsPath, "VC", "Tools", "MSVC");
            var toolsDir = Path.Combine(
                Directory.EnumerateDirectories(toolsRootDir).Last(),
                "bin", "HostX86", "x86");
            ClExePath = Path.Combine(toolsDir, "cl.exe");
            DumpBinPath = Path.Combine(toolsDir, "dumpbin.exe");
        }

        static string Run(string workingDirectory, string exePath, params string[] arguments)
        {
            using (var process = new Process())
            {
                process.StartInfo = new ProcessStartInfo(exePath, string.Join(" ", arguments))
                {
                    CreateNoWindow = true,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    WorkingDirectory = workingDirectory,
                };
                process.Start();
                process.WaitForExit();
                if (process.ExitCode != 0)
                {
                    throw new InvalidOperationException(process.StandardError.ReadToEnd());
                }
                return process.StandardOutput.ReadToEnd();
            }
        }

        public static string Compile(string filename)
        {
            Run(Path.GetDirectoryName(filename), ClExePath, "/nologo", "/c", "/Os", filename);
            return Path.ChangeExtension(filename, "obj");
        }

        public static string Disassemble(string filename)
        {
            var lines = Run(Path.GetDirectoryName(filename), DumpBinPath, "/disasm:bytes", filename)
                .Split(new [] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(l => l.Trim())
                .SkipWhile(l => l != "File Type: COFF OBJECT")
                .Skip(2) // Skip file type and signature
                .TakeWhile(l => l != "Summary");
            return string.Join("\n", lines);
        }

        public static TempDirectory CreateTempDirectory()
        {
            var tempFileName = Path.GetTempFileName();
            File.Delete(tempFileName);
            return new TempDirectory(tempFileName);
        }
    }
}

File: src/UnwindMC.Tests/SourceTests/VcTools.fs
module rec VcTools

open System
open System.Diagnostics
open System.IO
open System.Linq
open Newtonsoft.Json
open Newtonsoft.Json.Linq

let private findVcToolsPath(): string =
    let vswhereExePath =
        Path.Combine(
            Environment.GetEnvironmentVariable("ProgramFiles(x86)"),
            "Microsoft Visual Studio", "Installer", "vswhere.exe")
    let json = run null vswhereExePath [| "-format"; "json" |]
    let vsPath =
        JsonConvert
            .DeserializeObject<JArray>(json)
            .FirstOrDefault(fun o -> o.["installationVersion"].ToObject<string>().StartsWith("15."))
            .["installationPath"]
            .ToObject<string>()
    let toolsRootDir = Path.Combine(vsPath, "VC", "Tools", "MSVC")
    Path.Combine(
        Directory.EnumerateDirectories(toolsRootDir).Last(),
        "bin", "HostX86", "x86")

let private vcToolsPath = findVcToolsPath()
let private ClExePath = Path.Combine(vcToolsPath, "cl.exe")
let private DumpBinPath = Path.Combine(vcToolsPath, "dumpbin.exe")

let private run (workingDirectory: string) (exePath: string) (arguments: string array): string =
    use proc = new Process()
    proc.StartInfo <- new ProcessStartInfo(exePath, System.String.Join(" ", arguments))
    proc.StartInfo.CreateNoWindow <- true
    proc.StartInfo.RedirectStandardOutput <- true
    proc.StartInfo.RedirectStandardError <- true
    proc.StartInfo.UseShellExecute <- false
    proc.StartInfo.WorkingDirectory <- workingDirectory
    proc.Start() |> ignore
    proc.WaitForExit()
    if proc.ExitCode <> 0 then
        failwith (proc.StandardError.ReadToEnd())
    proc.StandardOutput.ReadToEnd()

let compile (filename: string): string =
    run (Path.GetDirectoryName(filename)) ClExePath [| "/nologo"; "/c"; "/Os"; filename |] |> ignore
    Path.ChangeExtension(filename, "obj")

let disassemble (filename: string): string =
    let lines =
        (run (Path.GetDirectoryName(filename)) DumpBinPath [| "/disasm:bytes"; filename |])
            .Split([| '\r'; '\n' |], StringSplitOptions.RemoveEmptyEntries)
            .Select(fun l -> l.Trim())
            .SkipWhile(fun l -> l <> "File Type: COFF OBJECT")
            .Skip(2) // Skip file type and signature
            .TakeWhile(fun l -> l <> "Summary")
    System.String.Join("\n", lines)

let CreateTempDirectory(): TempDirectory.T =
    let tempFileName = Path.GetTempFileName()
    File.Delete(tempFileName)
    new TempDirectory.T(tempFileName)
